Update SameSite Metatype to Support Partitioned

volosied opened the pull request:
fixes #27796

volosied commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -269,3 +269,6 @@ http.headers.setIfMissing.desc=Specifies the header names and values that are se
 http.headers.remove=Headers to be removed from each response
 http.headers.remove.desc=Specifies the header names that are removed from each HTTP response. Any header name that is defined by using this attribute must not be empty. No header values are expected. Any header name that is defined by using this attribute must not be present in the 'add', 'set', or 'setIfMissing' header configurations.
+
+http.samesite.partitioned=Parition all listed samesite none cookies. 
+http.samesite.partitioned.desc=Specifies that the cookies listed samesite none are to be partitioned. The default value is false meaning that cookies are not partitioned. 
```
This will need some work to get the wording right. 
```suggestion
http.samesite.partitioned=Partition all outgoing SameSite=None cookies. 
http.samesite.partitioned.desc=Specifies that all SameSite=None cookies are to be partitioned and therefore sent in cross-site requests. The default value is false meaning that cookies are not partitioned. 
```

Should I ignore SameSite=None mention and just say that the cookies are partitioned? Or only mention it in the description?

dmuelle replied:
```suggestion
http.samesite.partitioned.desc= When this attribute is set to true, all SameSite=None cookies are partitioned and therefore are sent in cross-site requests. The default value is false, which means that cookies are not partitioned. 
```

dmuelle reviewed (commented):
ID review- if is premature and you're still working out the wording, feel free to disregard this and just ping me whenever its ready for review. Thanks

volosied commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2004, 2023 IBM Corporation and others.
+ * Copyright (c) 2004, 202 IBM Corporation and others.
```
fix

volosied commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpChannelConfig.java:
```diff
@@ -194,6 +194,9 @@ public class HttpChannelConfig {
     private Map<Pattern, String> sameSitePatterns = null;
     private boolean onlySameSiteStar = false;
+    
+    /* Identifies if the partitioned cookie attrbute should be set */
```
spelling

volosied commented on dev/com.ibm.ws.transport.http/src/com/ibm/wsspi/http/HttpCookie.java:
```diff
@@ -230,6 +231,25 @@ public void setDiscard(boolean flag) {
         this.myDiscardFlag = flag;
     }
 
+    // /**
+    //  * Set the Partitioned attribute flag.
+    //  *
+    //  * @param flag
+    //  */
+    // public void setPartitioned(boolean flag) {
+    //     this.myPartitionedFlag = flag;
+    // }
+
+    // /**
+    //  * Query if the Partitioned attribute is set.
+    //  *
+    //  * @return boolean
+    //  */
+    // public boolean isPartitioned() {
+    //     return this.myPartitionedFlag;
+    // }
+
+
```
No longer needed since, we use setAttribute

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpChannelConfig.java:
```diff
@@ -1251,6 +1258,25 @@ private void parseCookiesSameSiteStrict(Map<Object, Object> props) {
     }
 
+    private void parseCookiesSameSitePartitioned(Map<Object, Object> props) {
+        System.out.println("parseCookiesSameSitePartitioned");
```
Let's make sure to remove system.out before merging

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/values/CookiePartitioned.java:
```diff
@@ -0,0 +1,62 @@
+    private static final TraceComponent tc = Tr.register(CookieSecureData.class, HttpMessages.HTTP_TRACE_NAME, HttpMessages.HTTP_BUNDLE);
+
+    /**
+     * Constructor for a Cookie's secure flag.
```
not for the secure flag :)

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/values/CookiePartitioned.java:
```diff
@@ -0,0 +1,62 @@
+        // cookie.setPartitioned(true);
+        cookie.setAttribute("partitioned", "");
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(tc, "Cookie HttpOnly attribute set");
```
Update to reflect partitioned

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/publish/servers/servlet40_partitioned/jvm.options:
```diff
@@ -0,0 +1 @@
+-Dcom.ibm.ws.beta.edition=true
```
new line

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/test-applications/PartitionedTest.war/resources/PartitionedSetCookie.jsp:
```diff
@@ -0,0 +1,43 @@
+        %>
+
+    </body>
+</html>
```
new line

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/FATSuite.java:
```diff
@@ -73,7 +74,7 @@
 @RunWith(Suite.class)
 @SuiteClasses({
-                // WCPushBuilderSecurityTest.class,
+                WCPushBuilderSecurityTest.class,
```
I'm assuming uncommenting this was accidental?

volosied replied:
Verify in integration -- may be commented out already?

volosied replied:
comment out

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/internal/HttpChain.java:
```diff
@@ -456,6 +457,10 @@ public synchronized void update(String resolvedHostName) {
                             enableSameSite = true;
                             chanProps.put(HttpConfigConstants.PROPNAME_SAMESITE_STRICT, samesiteOptions.get("strict"));
                         }
+                        if (samesiteOptions.containsKey("partitioned")) {
```
Do we need a test where we just set "partitioned=true" in the server.xml and then use the Servlet APIs to add samesite?

volosied replied:
Yes, that's a good test case. That scenario may not work with the current code, but I'll test it out.  It will need to be added in the 6.1 FAT. I have a few tests planned interactions with Cookie#setAttribute behavior.

pnicolucci replied:
Was a test added?

pnicolucci replied:
Also why would Servlet 6.1 FAT be required?

volosied replied:
Typo -- I mean servlet 6.0 fat (which I add tests too).

volosied replied:
Yes -  6.0 FAT: testSetAttributeWithPartitionedConfig

https://github.com/OpenLiberty/open-liberty/pull/27809/files/0ab5adce5a8f6f4d0a83f18d02d8cec49ad3cdc5#diff-f5ca25689c46ddee6312a8127f6d541acd8db6ade26af722bd7f0f52313c513fR135

volosied committed:
Implement Partitioned in Http Transport

Fix edge case SameSite=None Programmatic, but Paritioned set via Config

Fix non beta behavior (extra partitioned)

volosied committed:
Create Partitioned Tests in Servlet 4.0 FAT

volosied committed:
Create Paritioned Tests in Servlet 6.0 FAT

Add two new tests to CookieSetAttributeServlet: testSetAttributePartitioned & testNullValueRemovesAttribute

volosied commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/values/CookieData.java:
```diff
@@ -57,6 +57,9 @@ public abstract class CookieData extends GenericKeys {
     /** SameSite attribute */
     public static final CookieData cookieSameSite = new CookieSameSiteData();
+    /** Partitioned attribute */
+    // BETA - KeyMatcher will only Added Paritioned if in Beta
```
Reference KeyMatcher.java#add in comment

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat

volosied committed:
Fix unit test build issues (caused by beta checks)

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat

volosied committed:
Mark http.samesite.partitioned as internal instead

volosied commented on dev/com.ibm.ws.transport.http/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -269,3 +269,6 @@ http.headers.setIfMissing.desc=Specifies the header names and values that are se
 http.headers.remove=Headers to be removed from each response
 http.headers.remove.desc=Specifies the header names that are removed from each HTTP response. Any header name that is defined by using this attribute must not be empty. No header values are expected. Any header name that is defined by using this attribute must not be present in the 'add', 'set', or 'setIfMissing' header configurations.
+
+http.samesite.partitioned=Parition all listed samesite none cookies. 
+http.samesite.partitioned.desc=When this attribute is set to true, all SameSite=None cookies are partitioned and therefore are sent in cross-site requests. The default value is false, which means that cookies are not partitioned.
```
Can this still be included even if the metatype is now internal?

pnicolucci replied:
Sure, we just need to update the metatype.xml once we're ready to beta to point to the translations.

volosied commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/CookieUtils.java:
```diff
@@ -578,7 +600,12 @@ private static void setAttributes(HttpCookie cookie, StringBuilder buffer) {
             for (Entry<String, String> entry : cookieAttrs.entrySet()) {
                 key = entry.getKey();
-                if (!(key.equals("samesite") || key.equals("port") || key.equals("commenturl"))) {
+                // add partitioned if not in beta.
+                // skip if in beta since it's added above already
+                if((key.equals("partitioned") && ProductInfo.getBetaEdition())){
+                    continue;
+                }
+                if (!(key.equals("samesite") || key.equals("port") || key.equals("commenturl")) ) {
```
This part is tested by tested by Servlet60CookieSetAttributeTest's  `test_setAttributePartitioned` -- we expected `partitioned=` to be set if not in beta; otherwise, it's just `; Partitioned`

pnicolucci replied:
Not sure I'm following here, are we making a change that would work outside of Beta in a GA release?

volosied replied:
This change and test (test_setAttributePartitioned) I created are to verify that my Partitioned work is only in beta.

Currently, setting partitioned on the cookie renders "partitioned=". With my change, we would only render "Partitioned" -- no equals.

volosied commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/test-applications/PartitionedTest.war/src/partitioned/servlets/PartitionedSplitCookieServlet.java:
```diff
@@ -0,0 +1,53 @@
+
+            resp.setHeader("Set-Cookie", "PartitionedCookieName_SetHeader=PartitionedCookieValue_SetHeader; Partitioned");
+
+            /*
+             * Lax and Partitioned is not a proper configuration, but this test makes sure it still work (not prevented);
+             * Warning could be logged, but that would be verbose if this method is called often.
+             * Developers responsibility to assure proper attributes are set 
+             */
```
Remove comment -- not needed anymore

pnicolucci replied:
If not needed lets remove it :)

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java:
```diff
@@ -2921,12 +2921,32 @@ else if (getServiceContext().getHttpConfig().onlySameSiteStar()) {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "[2] Setting the Partitioned attribute for SameSite=None");
+                    }
+                    cookie.setAttribute("partitioned", "");
```
Just to verify this is the scenario where SameSite would have been set with setAttribute in the Servlet API or using set/addHeader in the Servlet API and defining a set-cookie header with samesite?

volosied replied:
this was added to for you scenario you gave:  `Do we need a test where we just set "partitioned=true" in the server.xml and then use the Servlet APIs to add samesite?`

volosied replied:
The first check (line 1925) looks if partitioned and samesite=none are set on **the config.** 

This second check (2941)  only looks if the samesite config is used and **the cookie** has samesite=none.   If so, we then check for partitioned and then add it if needed.

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpChannelConfig.java:
```diff
@@ -2984,6 +3012,13 @@ public boolean onlySameSiteStar() {
     }
 
+    /*
+     * Returns a boolean which indicates whether Partitioned should be added the the SameSite=None cookies. 
```
should be added `the the`

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/values/CookiePartitioned.java:
```diff
@@ -0,0 +1,60 @@
+public class CookiePartitioned extends CookieData {
+
+    /** Trace component for debugging */
+    private static final TraceComponent tc = Tr.register(CookieSecureData.class, HttpMessages.HTTP_TRACE_NAME, HttpMessages.HTTP_BUNDLE);
```
We should be using CookiePartitioned.class here unless I'm missing something.

Also should we stay consistent and do `CookiePartitionedData` like we did with `CookieSameSiteData`?

volosied replied:
Copy Paste error -- fixing

pnicolucci replied:
I see the naming convention is a bit all over some CookieX and other CookieXData, just planning to stick with CookiePartitioned?

volosied replied:
Yes, at this point, I'd rather stick with CookiePartitioned. We can refactor this code (and related classes) later on.

pnicolucci commented on dev/com.ibm.ws.webcontainer.cors/bnd.bnd:
```diff
@@ -51,7 +51,9 @@ instrument.classesExcludes: com/ibm/ws/webcontainer/cors/internal/resources/*.cl
 	com.ibm.ws.webcontainer;version=latest,\
 	com.ibm.ws.org.osgi.annotation.versioning;version=latest,\
-	com.ibm.ws.transport.http
+	com.ibm.ws.transport.http,\
```
Why are the changes to cors necessary?

volosied replied:
Unittests in cors use com.ibm.ws.transport.http code which now contains beta checks.   These beta classes weren't part of the classpath and caused ClassNotFoundExceptions

pnicolucci replied:
ok can you create an issue to clean this up once we're not beta any longer?

volosied replied:
Yep: https://github.com/OpenLiberty/open-liberty/issues/28070

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0/bnd.bnd:
```diff
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2017, 2023 IBM Corporation and others.
+# Copyright (c) 2017, 2024gi IBM Corporation and others.
```
`gi`

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCPartitionedAttributeTests.java:
```diff
@@ -0,0 +1,570 @@
+        ShrinkHelper.defaultDropinApp(server, APP_NAME + ".war", "partitioned.servlets");
+
+        // Start the server and use the class name so we can find logs easily.
+        server.startServer(WCSameSiteCookieAttributeTests.class.getSimpleName() + ".log");
```
`WCPartitionedAttributeTests`

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCPartitionedAttributeTests.java:
```diff
@@ -0,0 +1,570 @@
+@RunWith(FATRunner.class)
+public class WCPartitionedAttributeTests {
+
+    private static final Logger LOG = Logger.getLogger(WCSameSiteCookieAttributeTests.class.getName());
```
`WCPartitionedAttributeTests`

volosied committed:
Clean up / review comments

volosied commented:
#build

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpChannelConfig.java:
```diff
@@ -2984,6 +3012,13 @@ public boolean onlySameSiteStar() {
     }
 
+    /*
+     * Returns a boolean which indicates whether Partitioned should be added the SameSite=None cookies. 
```
`should be added to the SameSite=None cookies.`

volosied replied:
Will fix in the next PR

pnicolucci reviewed (approved):
I'm ok with you making the minor outstanding changes in a new pr so that we can incrementally make progress!

volosied commented:
#libby
